tests/integration/test_kfp_functional.py

```diff
@@ -0,0 +1,218 @@
+        raise_on_error=True,
+        timeout=1800,
+    )
+
+    # Wait for kfp-* Deployment to be active and idle.
+    # If a kfp charm is ever migrated to sidecar, and the above issue is still
+    # there, appropriate changes must be done on this method for checking the readiness
+    # of the application.
+    # FIXME: This is a workaround for issue https://bugs.launchpad.net/juju/+bug/1981833
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=2, min=1, max=15),
+        stop=tenacity.stop_after_delay(900),
+        reraise=True,
+    )
+    def assert_get_kfp_deployment(kfp_component: str):
+        """Asserts the deployment of kfp_component is ready and available."""
+        log.info(f"Waiting for {kfp_component} to be ready and available")
+        # Get Deployment, the namespace should always be kubeflow
+        kfp_deployment = lightkube_client.get(Deployment, name=kfp_component, namespace="kubeflow")
+
+        # Verify the deployment exists
+        assert kfp_deployment is not None
+
+        # Verify the readiness and availability
+        assert kfp_deployment.status.readyReplicas == 1
+        assert kfp_deployment.status.availableReplicas == 1
+
+    for component in KFP_CHARMS:
+        if component == "kfp-api":
+            # The issue seems to be happening only with podspec charms,
+            # nothing to do for kfp-api
+            continue
+        assert_get_kfp_deployment(kfp_component=component)
+
+
+# ---- KFP API Server focused test cases
+async def test_upload_pipeline(kfp_client):
+    """Upload a pipeline from a YAML file and assert its presence."""
+    # Upload a pipeline and get the server response
+    pipeline_upload_response = kfp_client.pipeline_uploads.upload_pipeline(
+        uploadfile=SAMPLE_PIPELINE, name="test-upload-pipeline"
+    )
+    # Upload a pipeline and get its ID
+    uploaded_pipeline_id = pipeline_upload_response.id
+
+    # Get pipeline id by name, default='sample-pipeline'
+    server_pipeline_id = kfp_client.get_pipeline_id(name="test-upload-pipeline")
+    assert uploaded_pipeline_id == server_pipeline_id
+
+
+async def test_create_and_monitor_run(kfp_client, create_and_clean_experiment):
+    """Create a run and monitor it to completion."""
+    # Create a run, save response in variable for easy manipulation
+    # Create an experiment for this run
+    experiment_response = create_and_clean_experiment
+
+    # Create a run from a pipeline file (SAMPLE_PIPELINE) and an experiment (create_experiment).
+    # This call uses the 'default' kubeflow service account to be able to edit Workflows
+    create_run_response = kfp_client.create_run_from_pipeline_package(
+        pipeline_file=SAMPLE_PIPELINE,
+        arguments={},
+        run_name="test-run-1",
+        experiment_name=experiment_response.name,
+        namespace=KUBEFLOW_PROFILE_NAMESPACE,
+    )
+
+    # FIXME: waiting_for_run_completion timeouts on GitHub runners
```

Comment:
I have added an issue: https://github.com/canonical/kfp-operators/issues/244

Reply:
What state is this run and its k8s objects in when it is hanging?  Is this something that would eventually resolve (eg: just needs more time/resources) or something that is broken?  I get the feeling you're saying it just needs more time/resources, but these same symptoms could be because it just isn't running the job properly

Reply:
It is running, it will have a state, but it doesn't seem to be able to finish within the timeout limits. If I increase the timeout, I then hit another problem with the `port-forward` losing connection to the svc it is forwarding.

What I will do is retrieve that state, check that it is running at least until we figure out a solution for the issue. Would that be okay?

Reply:
My worry is that it is not finishing because of a configuration problem within kfp-api.  I think we can misconfigure kfp-api in a way where it can start jobs, but cannot keep up to date on their progress (so they'll perpetually look like they're running).  I don't think the runner is simply overloaded here, I think what you're seeing is your tests correctly showing you a real bug

Reply:
I'm curious what the Pods, deployments, etc from the jobs say (`kubectl get` and `kubectl describe` on pod, deployment, and all the pipelines-related objects for a run)

Reply:
@DnPlas since we have a follow-up issue, could we expose some logs from Pods and add them on the other issue and unblock this PR?

While I'd also prefer that we have the answer on why it times out only in the GH Runner at the same time let's continue with merging this first PR.

Reply:
@dnplas and I discussed.  Agreed that this should be addressed as a separate issue

Reply:
@kimwnasptd there is a GH issue for tracking this #244, I will be adding logs there. @ca-scribner and I have discussed this offline and have confirmed that this behaviour is only observed on the GH runners, when running on a machine with more resources, the `Run` actually gets `status == 'Succeeded'`.
I will resolve this conversation, as the issue should cover all details, but feel free to re-open if further discussion is needed.